src/lib.rs

```diff
@@ -307,6 +314,88 @@ impl Default for Options {
     }
 }
 
+#[derive(Debug)]
+/// A raw image definition which can be used to create an optimized png
+pub struct RawImage {
+    png: Arc<PngImage>,
+}
+
+impl RawImage {
+    /// Construct a new raw image definition
+    ///
+    /// * `width` - The width of the image in pixels
+    /// * `height` - The height of the image in pixels
+    /// * `color_type` - The color type of the image
+    /// * `bit_depth` - The bit depth of the image
+    /// * `data` - The raw pixel data of the image
+    pub fn new(
+        width: u32,
+        height: u32,
+        color_type: ColorType,
+        bit_depth: BitDepth,
+        data: Vec<u8>,
+    ) -> Result<Self, PngError> {
+        // Validate bit depth
+        let valid_depth = match color_type {
+            ColorType::Grayscale { .. } => true,
+            ColorType::Indexed { .. } => (bit_depth as u8) <= 8,
+            _ => (bit_depth as u8) >= 8,
+        };
+        if !valid_depth {
+            return Err(PngError::InvalidDepthForType(bit_depth, color_type));
+        }
+
+        // Validate data length
+        let bpp = bit_depth as usize * color_type.channels_per_pixel() as usize;
+        let row_bytes = (bpp * width as usize + 7) / 8;
+        let expected_len = row_bytes * height as usize;
+        if data.len() != expected_len {
+            return Err(PngError::IncorrectDataLength(data.len(), expected_len));
+        }
+
+        Ok(Self {
+            png: Arc::new(PngImage {
+                ihdr: IhdrData {
+                    width,
+                    height,
+                    color_type,
+                    bit_depth,
+                    interlaced: Interlacing::None,
+                },
+                data,
+                aux_headers: IndexMap::new(),
+            }),
+        })
+    }
+
+    /// Add a png header chunk, such as "iTXt", to be included in the output
+    pub fn add_png_header(&mut self, chunk_type: [u8; 4], data: Vec<u8>) {
```

Comment:
Header is not a good name, since in PNG iHDR is the header. Everything is a chunk.

Is correct order of chunks enforced somewhere? Some must be before pLTE or before iDAT.

Reply:
Good point, changed.
And yes, the output function enforces certain chunks before/after the PLTE.